src/pagination/index.tsx

```diff
@@ -84,105 +109,167 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
 
 	if (pageSizes !== undefined && pageSizes !== icache.get('pageSizes')) {
 		icache.set('pageSizes', pageSizes);
-		icache.set('pageSizesData', pageSizes.map((ps) => ({ value: ps.toString() })));
+		icache.set(
+			'pageSizesResource',
+			createMemoryResourceWithData(pageSizes.map((ps) => ({ value: ps.toString() })))
+		);
 	}
 
 	const page = icache.getOrSet('currentPage', 1);
 	const pageSize = icache.getOrSet('pageSize', 10);
-	const pageSizesData = icache.get('pageSizesData');
+	const pageSizesResource = icache.get('pageSizesResource');
+
+	const getRenderedWidth = (dnode: RenderResult) => {
+		if (dnode === undefined) {
+			return 0;
+		}
+
+		const r = renderer(() => dnode);
+		const div = global.document.createElement('div');
+		div.className = theme.variant() || '';
+		div.style.position = 'absolute';
+		global.document.body.appendChild(div);
+		r.mount({ domNode: div, sync: true });
+		const dimensions = div.getBoundingClientRect();
+		global.document.body.removeChild(div);
+		return dimensions.width;
+	};
+
+	// If provided, use `siblingCount` as an upper bound for sibling links
+	const leadingCount = siblingCount ? Math.min(siblingCount, page - 1) : page - 1;
+	const trailingCount = siblingCount ? Math.min(siblingCount, total - page) : total - page;
+
+	const prevLink = leadingCount ? (
+		<button
+			type="button"
+			key="prev"
+			classes={[classes.prev, classes.link]}
+			onclick={(e) => {
+				e.stopPropagation();
+				icache.set('currentPage', page - 1);
+				onPage && onPage(page - 1);
+			}}
+		>
+			<div classes={classes.icon}>
+				<Icon type="leftIcon" />
+			</div>
+			<div classes={classes.label}>{messages.previous}</div>
+		</button>
+	) : (
+		undefined
+	);
+
+	const nextLink = trailingCount ? (
+		<button
+			type="button"
+			key="next"
+			classes={[classes.next, classes.link]}
+			onclick={(e) => {
+				e.stopPropagation();
+				icache.set('currentPage', page + 1);
+				onPage && onPage(page + 1);
+			}}
+		>
+			<div classes={classes.icon}>
+				<Icon type="rightIcon" />
+			</div>
+			<div classes={classes.label}>{messages.next}</div>
+		</button>
+	) : (
+		undefined
+	);
+
+	const leadingSiblings: RenderResult[] = [];
+	const trailingSiblings: RenderResult[] = [];
+	const containerWidth = dimensions.get('links').size.width;
+	let containerStyles: Partial<CSSStyleDeclaration> = {
+		opacity: containerWidth ? '1' : '0'
+	};
+
+	const current = (
+		<div key="current" classes={classes.currentPage}>
+			{page.toString()}
+		</div>
+	);
 
-	const leading: WNode[] = [];
-	const leadingCount = Math.min(siblingCount, page - 1);
-	for (let i = 1; i <= leadingCount; i++) {
-		const linkTo = page - i;
+	let availableWidth =
+		containerWidth -
+		getRenderedWidth(prevLink) -
+		getRenderedWidth(current) -
+		getRenderedWidth(nextLink);
 
-		leading.unshift(
+	for (let i = 1; i <= Math.max(leadingCount, trailingCount); i++) {
```

Comment:
Feels to me that this should just calculate what there's room for and render it, this trial and error approach will be costly. Should just calc and store the width of elements without buttons, and the width of a button then on resize, calculate how many fill fit, perhaps use a threshold of 90% available space? When calculating the width of the button up front you could use the largest page number as it's label, then you can always be sure they'll fit.